### DISCUSSION TITLE
User interface design considerations.

### DISCUSSION THREAD

**Discussion by macbugs** [textbrowser/spot-on#14] [Ideas]
Technically speaking, it appears to be a capable platform, but the UI is rather complex.  Some of this complexity could be represented by icons or controls and hidden behind the main panel somewhere.  I guess the term is _“easy onboarding”_ — you dont want to scare the user with difficult choices (or a “busy” user interface.)

The general public is going to balk when confronted with the need to choose a cipher, for example.  I think maybe the default should be the strongest protection, and the cipher parameters should be hidden behind a switch, so the option to change them only appears if the default is not ‘activated’.  Why should the user choose anything but the strongest settings?  Because key generation takes too long on a slow CPU?  (And how long?)  ...It appears that the key generation process is only tasking one CPU core.  No problem here, its fast enough.  But how long does it take to generate the strongest private key on some old CPU?

—Just curious about the reason for presenting such complex choices to the ordinary user.  If you enforce minimum password length, you are making security decisions for the user.  No problem — the restriction is for a good reason.  But perhaps you should do the same with cipher parameters (or at least indicate the criteria for choosing different parameters, so the user can make an informed decision.)  You cannot expect him to know which cipher/key size is appropriate.  (He does not know why he should choose one or the other.)
___

SSL control strings on the main panel of the UI ?

![Screenshot 2022-11-30 at 08 46 54](https://user-images.githubusercontent.com/101867641/204839244-c4f6c8c7-9f32-4489-a77a-6ac64f951480.png)

And the UI has a few glitches… (does it actually look like this on other platforms?)

![Screenshot 2022-11-30 at 08 36 35](https://user-images.githubusercontent.com/101867641/204839419-731a34df-7b3b-4a95-b81b-d400f05cd0c0.png)

Was something supposed to appear in this rectangle?

![Screenshot 2022-11-30 at 08 44 36](https://user-images.githubusercontent.com/101867641/204839471-19932ce5-4182-49d2-8ef4-3569f5aa43ff.png)

![Screenshot 2022-11-30 at 08 50 27](https://user-images.githubusercontent.com/101867641/204839508-9e9e94b4-fc6e-445a-bd73-e655398747df.png)

**Comment by textbrowser** [comment]
The tab icon size is configurable. I'll change the default size from 24, 24 to 16, 16.

![Screenshot_20221130_120018](https://user-images.githubusercontent.com/10701156/204860965-c2a48df8-ee66-40b8-aea4-55d49760d536.png)

**Comment by textbrowser** [comment]
The elided text is normal too. You can hide some pages.

**Comment by textbrowser** [comment]
GCrypt's key-generation functions are not interruptible. I suppose the processes could be performed in other tasks. I don't know if doing so would be safe. It would add uncomfortable complexity to the development with respect to thread-safety and process termination. NTRU is not interruptible. I think I could make McEliece interruptible but not NTL which McE relies on.

**Comment by textbrowser** [comment]
Can you include complete pictures? I'll post my Mac views so you can see what it looks like.

**Comment by macbugs** [comment]
![Screenshot 2022-11-30 at 12 47 29](https://user-images.githubusercontent.com/101867641/204870707-de087338-8996-45e9-9cc6-a97e98f51e8e.png)
![Screenshot 2022-11-30 at 12 49 10](https://user-images.githubusercontent.com/101867641/204871109-351c43b0-80af-4553-9bdc-0a4a467c3f7c.png)

**Reply by textbrowser** [reply]
Funky!

**Reply by macbugs** [reply]
[🪩](https://www.youtube.com/watch?v=gjKFCYzqq-A)

**Comment by macbugs** [comment]
This is full-screen view with the default setting:

![Screenshot 2022-11-30 at 13 08 59](https://user-images.githubusercontent.com/101867641/204875096-d2aba9f2-5f9c-4c4a-ad16-f52257fb7de8.png)

> **_The tab icon size is configurable. I'll change the default size from 24 to 16_**

Yes, that seems to work.  (And maybe tab string should be centered on tab?)
![Screenshot 2022-11-30 at 13 19 55](https://user-images.githubusercontent.com/101867641/204877356-1d6cef40-5f05-492e-9a1a-cb73056521c3.png)
...but I think you should probably enforce some minimum window size that prevents this label truncation

**Comment by textbrowser** [comment]
Testing on Mac is prohibitive because my old hardware, lack of interest, and the unfriendliness of the environment. Without Homebrew, I would be tempted to abandon it.

**Comment by textbrowser** [comment]
I don't understand those gaps beneath the buttons. I don't see them. Let me test a new build with a dark theme.

**Comment by textbrowser** [comment]
The tab widgets probably require tweaks. Perhaps document mode.

**Comment by textbrowser** [comment]
As for the complexity of options, there is an incomplete brief version: View -> Minimal Display. It's super incomplete. :)

**Comment by textbrowser** [comment]
And some of your tab headers are a bit transparent, right?

**Comment by textbrowser** [comment]
![so-3](https://user-images.githubusercontent.com/10701156/204887007-079f05a9-e141-4779-ae3d-32143ad170c7.png)
![so-2](https://user-images.githubusercontent.com/10701156/204887013-2e944f4a-2b0b-4dba-83aa-f4e01a07abe2.png)
![so-1](https://user-images.githubusercontent.com/10701156/204887015-7b05c381-a39c-450b-b826-be20e5ceea2e.png)

**Reply by textbrowser** [reply]
I'll apply document mode to all other tab widgets.

**Comment by textbrowser** [comment]
What would you consider a reasonable minimum password length? It's simple to modify it.

**Comment by textbrowser** [comment]
As for default ciphers and digests, the selections are based on popularity. I leave safety and strength to the individual.

Which of the options should be hidden? I'll make some selections and publish those.

**Comment by textbrowser** [comment]
I like configurable things, including software. Hidden variables bother me more than complex concepts so that's the reason there are so many widgets.

**Comment by macbugs** [comment]
> **_And some of your tab headers are a bit transparent, right?_**

No, the dark theme is just a system theme applied to all apps which support it.  Transparency is not an option.

![Screenshot 2022-12-01 at 08 13 06](https://user-images.githubusercontent.com/101867641/205061754-4b81847f-99aa-4354-955f-8e5939cf8a9f.png)

> **_What would you consider a reasonable minimum password length?_**

Most people are accustomed to memorizing a 4-digit code for bank cards.  Maybe later, could support Bluetooth device proximity unlock or Yubikey type device.  Take "[Wire.app](https://github.com/wireapp/wire)" for example: the user can choose to save the password, so he does not have to type it every time the app launches.  Application passwords can be saved in the system keychain.

> **_I leave safety and strength to the individual._**

It is nice to have a choice... but the choice confuses the user.  It should be hidden in some "advanced" panel, like VLC:

![image](https://user-images.githubusercontent.com/101867641/205072374-abb33ca9-aca6-471a-aac1-43c71ba4a2c3.png)

The cipher choice intimidates the typical user.  If I cannot convince my contacts to use alternative messaging platforms like Spot-On, they will stick with e-mail & SMS.

> **_the reason there are so many widgets_**

The public demands simplicity, or it will reject the app, and continue to use something insecure.  Signal app is popular because it is easy.  It does not satisfy me -- but if my contacts will use it, that is better than _**no**_ security.  Configuration options should be moved to another window or panel (should not be displayed on the main user interface).

**Reply by textbrowser** [reply]
One of the images you published, the E-Mail page... do you see the C/O tab header? It's a wee bit transparent. I mean, you can see a horizontal line behind the tab. Is it a Qt-ism or Mac-ism? Anyway, perhaps the document mode will also work better for you.

Four characters are fine.

A minimal view is for the future.

**Reply by macbugs** [reply]
ok, I see it now... I think that is the fault of Qt... because of the spurious rectangles on the same line.  It could be rendering text, some kind of padding somewhere.
![Screenshot 2022-12-01 at 11 35 51](https://user-images.githubusercontent.com/101867641/205108527-de9a91d6-145a-45e2-8641-9c8873991fc0.png)
btw, here is RetroShare on Mac, whoopee! 🎉 😏
```
Exception Type:        EXC_CRASH (SIGABRT)
Exception Codes:       0x0000000000000000, 0x0000000000000000

Termination Reason:    Namespace DYLD, Code 1 Library missing
Library not loaded: /usr/local/Cellar/nettle/3.4.1/lib/libnettle.6.dylib
Referenced from: <8A53F23D-4D30-39C8-8C26-D10A8960A309> /Volumes/VOLUME/*/retroshare.app/Contents/Frameworks/libhogweed.4.dylib
Reason: tried: '/usr/local/Cellar/nettle/3.4.1/lib/libnettle.6.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/usr/local/Cellar/nettle/3.4.1/lib/libnettle.6.dylib' (no such file), '/usr/local/Cellar/nettle/3.4.1/lib/libnettle.6.dylib' (no such file), '/usr/local/lib/libnettle.6.dylib' (no such file), '/usr/lib/libnettle.6.dylib' (no such file, not in dyld cache)
(terminated at launch; ignore backtrace)
```

**Reply by textbrowser** [reply]
Moving aside /usr/local/Cellar will prevent the local program from launching. Also reviewing all open files.

**Comment by textbrowser** [comment]
I view simplicity as lack of power. Power belongs to you and software which allows you the power will be unusually complex.

**Comment by textbrowser** [comment]
I also agree with you. :)

**Comment by textbrowser** [comment]
New minimum password length committed!
https://github.com/textbrowser/spot-on/commit/a9b892141d5e68e6b352ba1afcf81454d3bc9381

**Comment by macbugs** [comment]
[_“Similar to the cockpit of an aircraft”_](https://compendio.github.io/goldbug-manual/) 😂

--Some design concepts for setup wizard / password & key management...
![Screenshot 2022-12-02 at 08 16 28](https://user-images.githubusercontent.com/101867641/205446434-78845e83-a768-4549-a3c9-d347297ce2d6.png)
![mac-screenshot-1](https://user-images.githubusercontent.com/101867641/205446443-b0f02a09-c653-434a-a2bf-9ce33e639f25.png)
![Screenshot 2022-12-02 at 07 44 21](https://user-images.githubusercontent.com/101867641/205446460-58e5153d-7dda-41ed-b75f-9d1c4d23b02e.png)
![Screenshot 2022-12-02 at 07 47 26](https://user-images.githubusercontent.com/101867641/205446471-d3c348b9-a255-4aa9-b2bb-82560f3e1a0c.png)

**Reply by textbrowser** [reply]
An installation wizard is included. :P

**Reply by macbugs** [reply]
Statistical studies of user behavior demonstrate that first impressions have a significant affect on the rate of "uptake"
(...in other words, a UI needs to have _style_.) [😉](https://www.youtube.com/watch?v=OPf0YbXqDm0)

**Reply by textbrowser** [reply]
Perhaps. Theming does not interest me. Styles require icons, backgrounds, portability, fonts, testing. Who will do that?

**Reply by textbrowser** [reply]
I can spend 30 months on themes alone that I'll never enjoy and that will never be enjoyed by anyone else. And then when Apple breaks my effort, what's the outcome? Android suffers from fragmentation. I could not reproduce your transparency problems on my Mac (dark and light mode). Mozilla has people theming out. Like that's all they do, theme out. Themes in the morning, lunch, night, dreaming in themes.

**Reply by textbrowser** [reply]
I'm not creating to attract interest. I find the programs important so I share them. I try for portability, documentation, correctness, accessibility, openness. I dabble in Android. It annoys me. Even with the most generic implementations, there is misbehavior in version and / or device. I will have one program which was fine on a device and which will then be different after I upgrade the Android version. So even if you specify the most primitive of functions, Android does not guarantee that those functions will remain consistent in the next version. Well, Qt suffers from distribution to distribution and sometimes versions introduce problems.

Simple and detailed views are important, let's complete those first.

**Reply by macbugs** [reply]
> _**I could not reproduce your transparency problems**_

I have reviewed hundreds of apps and the problem only seems to exist in Qt.  This would not occur in Flutter where cross-platform UI support is better.  (Just my opinion, not suggesting a refactor here.)

**Reply by textbrowser** [reply]
A rewrite would produce a new appearance.

**Reply by macbugs** [reply]
I meant that I was not suggesting a major change like switching to Flutter... and I have some concerns about the protocol efficiency on mobile platforms.

**Comment by macbugs** [comment]
> _**And then when Apple breaks my effort**_

Yes, Apple changes various API's perhaps too frequently -- but not stuff like this.

> **_Android suffers from fragmentation._**

Google has acknowledged this problem, and effort is being made to address that issue in the design of its _"next generation"_ Fuchsia-based mobile OS.  I am certain it will replace Android in a few years (but it will have a 'compatibility mode' for Android apps.)  Samsung also has a Fuchsia mobile OS in development.

> _**I dabble in Android. It annoys me.**_

Agree.

> **_Simple and detailed views are important, let's complete those first._**

Yes, of course -- it is not an urgent matter, just a general discussion about design philosophy

**Reply by textbrowser** [reply]
Style is not owned by Apple and Google. These are two companies that attempt to define acceptable art. If a process violates Apple and Google styles, it's not acceptable. That's nonsense. If a process has instructions, it's meaningful regardless of its complexity. The metrics regarding simplicity are driven by metrics for popularity and acceptance.

Literature which does not follow strict style thrives.

**Reply by macbugs** [reply]
> _**Style is not owned by Apple and Google. These are two companies that attempt to define acceptable art.**_

Yes, I understand... and even Apple [does not obey](https://scribe.froth.zone/@krvoller/how-iphone-violates-apples-accessibility-guidelines-6785172eb343) its own [design guidelines](https://developer.apple.com/design/human-interface-guidelines/platforms/designing-for-macos/#//apple_ref/doc/uid/TP30000894-TP6)...

...And too many people will only use SMS or iMessage because there is no ideal cross-platform solution.

**Reply by textbrowser** [reply]
Isn't that annoying? E-mail is cross-platform. In Spot-On, you can communicate through e-mail in Poptastic. It was difficult to implement because e-mail is difficult (as is CURL), but it works. So, it could be made to work. SMS? There is a version called MMS which could work with real-time e-mail messaging.

**Reply by textbrowser** [reply]
E-mail should be re-evaluated anyway. It requires too much parsing. It needs new eyes and ideas.

**Reply by textbrowser** [reply]
https://en.wikipedia.org/wiki/Multimedia_Messaging_Service

**Comment by textbrowser** [comment]
Apple and Google broke the sacred idea of not judging a book by its cover.

**Comment by textbrowser** [comment]
https://en.wikipedia.org/wiki/Monism

**Reply by macbugs** [reply]
[㆝](https://www.youtube.com/watch?v=Ly6ZhQVnVow)
